src/client/legacy/connect/http.rs

```diff
@@ -433,7 +433,12 @@ where
                 .map_err(ConnectError::dns)?;
             let addrs = addrs
                 .map(|mut addr| {
-                    addr.set_port(port);
+                    // Respect explicit ports in the URI,
+                    // and non `0` ports resolved from a custom dns resolver.
+                    if dst.port().is_some() || addr.port() == 0 {
```

Comment:
Because of the combination of logical conditionals, I think it'd be wise to extract this little bit of code into a function that we can write a few unit tests for inside this file, down at the bottom. Not tests going through the whole resolution step, just once an address is resolved, the combination of port missing, port explicitly in `dst`, and default scheme port. What do you think?

Reply:
Yes sure makes sense, I tested few combinations manually, and considered writing tests but the whole process was daunting to add tests too, testing a function for just this would be easy, will do.